src/gt4py/next/program_processors/codegens/gtfn/gtfn_ir.py

```diff
@@ -126,23 +124,43 @@ def _values_validator(
     ) -> None:
         if not all(
             isinstance(el, (SidFromScalar, SidComposite))
-            or _is_ref_literal_or_tuple_expr_of_ref(el)
+            or _is_tuple_expr_of(lambda expr: isinstance(expr, (SymRef, Literal)), el)
             for el in value
         ):
             raise ValueError(
-                "Only 'SymRef', tuple expr of 'SymRef', 'SidFromScalar', or 'SidComposite' allowed."
+                "Only 'SymRef', 'Literal', tuple expr thereof, 'SidFromScalar', or 'SidComposite' allowed."
             )
 
 
+def _might_be_scalar_expr(expr: Expr) -> bool:
+    if isinstance(expr, BinaryExpr):
+        return all(_is_tuple_expr_of(_might_be_scalar_expr, arg) for arg in (expr.lhs, expr.rhs))
+    if isinstance(expr, UnaryExpr):
+        return _is_tuple_expr_of(_might_be_scalar_expr, expr.expr)
+    if (
+        isinstance(expr, FunCall)
+        and isinstance(expr.fun, SymRef)
+        and expr.fun.id in ARITHMETIC_BUILTINS
+    ):
+        return all(_might_be_scalar_expr(arg) for arg in expr.args)
+    if isinstance(expr, CastExpr):
+        return _might_be_scalar_expr(expr.obj_expr)
+    if _is_tuple_expr_of(lambda e: isinstance(e, (SymRef, Literal)), expr):
+        return True
+    return False
+
+
 class SidFromScalar(Expr):
     arg: Expr
 
     @datamodels.validator("arg")
     def _arg_validator(
         self: datamodels.DataModelTP, attribute: datamodels.Attribute, value: Expr
     ) -> None:
-        if not _is_ref_literal_or_tuple_expr_of_ref(value):
-            raise ValueError("Only 'SymRef' or tuple expr of 'SymRef' allowed.")
+        if not _might_be_scalar_expr(value):
```

Comment:
I don't see the usefulness of this function, can you explain?

Reply:
I can not guarantee that the expression is a scalar, but at least I can try to catch obvious errors. This is what the function is doing.